db: schema_tables: Consider only affected tables when computing schema diff

tgrabiec opened the pull request:
Currently, applying a schema change on a replica works like this:

  1. Collect all affected keyspaces from incoming mutations
  2. Read current state of schema
  3. Apply the mutations
  4. Read new state of schema
 
The "Read ... state of schema" step reads all kinds of schema objects. In particular, to read the "table" objects, it does the following:

```
for every affected keyspace k:
     read all mutations from system_schema.tables for k
     extract all existing table names from those mutations
     for every existing table:
           read mutations from {tables, columns, indexes, view_virtual_columns, ...} for that table 
```

As you can see, the number of reads performed is O(nr tables in a keyspace), not O(nr tables in a change).

This means that making a sequence of schema changes like adding a table is quadratic.

Another aspect which magnifies this is that we don't read those tables using a single scan, but issue individual queries for each table separately.

This patch optimizes this by considering only affected tables when reading schema for the purpose of diff calculation.

I tested this using a test case provided by Kamil (https://github.com/kbr-scylla/scylladb/commit/53fe1543a727931001416243bb0f78f4f4662d2d)

      ./test.py --mode debug test_many_schema_changes -s 

The test bootstraps a cluster and then creates about 40 schema changes. Then a new node is bootstrapped and replays those changes via group0.

In debug mode, each change takes roughly 2s to process before the patch, and 0.5s after the patch.

The whole replay is reduced to 56% of what was before:

Before (1m19s) :  

    INFO  2023-01-20 19:44:35,848 [shard 0] raft_group0 - setup_group0: ensuring that the cluster has fully upgraded to use Raft...
    INFO  2023-01-20 19:45:54,844 [shard 0] raft_group0 - setup_group0: waiting for peers to synchronize state...

After (45s):

    INFO  2023-01-20 22:02:51,869 [shard 0] raft_group0 - setup_group0: ensuring that the cluster has fully upgraded to use Raft...
    INFO  2023-01-20 22:03:36,834 [shard 0] raft_group0 - setup_group0: waiting for peers to synchronize state...

tgrabiec commented:
Refs https://github.com/scylladb/scylladb/issues/12591 (only partially fixes).

Posting for review, not ready for merging due to lack of unit tests.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3941/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3942/

avikivity commented on db/schema_tables.cc:
```diff
@@ -974,23 +990,32 @@ static read_table_names_of_keyspace(distributed<service::storage_proxy>& proxy,
 future<std::map<table_id, schema_mutations>>
-read_tables_for_keyspaces(distributed<service::storage_proxy>& proxy, const std::set<sstring>& keyspace_names, schema_ptr s)
+read_tables_for_keyspaces(distributed<service::storage_proxy>& proxy, const std::set<sstring>& keyspace_names, schema_ptr s,
+                          const std::unordered_map<sstring, std::optional<std::set<sstring>>>& tables_per_keyspace)
```
Long-standing dislike about the fact that we read anything from disk. We have the entire schema in memory already.

avikivity replied:
Why `std::optional<std::set>`? Won't an empty set, or missing entry in the map suffice?

tgrabiec replied:
missing entry means we haven't proccessed anything for the keyspace yet, while std::nullopt means that we processed some mutation and figured out that the set of affected tables is potentially unbounded. In this case, we must not override std::nullopt with a more narrow set coming from another mutation. An empty set is no good too, because it means no tables are affected, while std::nullopt means that many are affected.

avikivity replied:
Yes, I saw it, but it's adding meaning to std::optional in tricky and rarely touched code. We should be more explicit.

A counterexample is std::optional used for NULL, but there the meaning is similar, and the code is widely used so it's reasonable to assume the reader is familiar with the convention.

avikivity commented on db/schema_tables.cc:
```diff
@@ -974,23 +990,32 @@ static read_table_names_of_keyspace(distributed<service::storage_proxy>& proxy,
-    for (auto&& keyspace_name : keyspace_names) {
-        for (auto&& table_name : co_await read_table_names_of_keyspace(proxy, keyspace_name, s)) {
+    for (auto&& [keyspace_name, maybe_tables] : tables_per_keyspace) {
+        for (auto&& table_name : *maybe_tables) {
```
No check for optional::has_value?

tgrabiec replied:
It's guaranteed to be engaged here. We read the missing maybe_tables in the caller earlier.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1120,95 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+// If returns std::nullopt, it means that the set of affected tables is unknown.
+// The mutation must target one of the tables in schema_tables_holding_schema_mutations().
+static
+std::optional<std::set<sstring>> get_affected_tables(const sstring& keyspace_name, const mutation& m) {
```
The std::optional is explained, but I think it's too tricky to use std::optional to represent all.

How about std::variant<all_tables, std::unordered_set>? It's verbose but I think verbosity is important for such tricky and critical code.

avikivity replied:
also: why std::set and not std::unordered_set?

tgrabiec replied:
ACK for both suggestions.

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1120,95 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+                    co_await read_table_names_of_keyspace(proxy, keyspace_name, tables()));
+            table_names.merge(boost::copy_range<std::set<sstring>>(
+                    co_await read_table_names_of_keyspace(proxy, keyspace_name, views())));
+            maybe_tables = std::move(table_names);
```
There's a bug here. If maybe_tables is std::nullopt, we should still collect known affected tables from the mutations and add them on top of what is read here. Otherwise we may miss new tables in the post-apply phase.

nyh commented:
@tgrabiec is this a regression related to new schema-changes code or old code?
Could this be the cause of the very old issue https://github.com/scylladb/scylladb/issues/7620 where I discovered that schema changes are not only very slow, and not only O(number of unrelated tables), they count of tables even includes previously-deleted tables?

tgrabiec commented:
> @tgrabiec is this a regression related to new schema-changes code or old code? 

No, it's not a regression.

> Could this be the cause of the very old issue #7620 where I discovered that schema changes are not only very slow, and not only O(number of unrelated tables), they count of tables even includes previously-deleted tables?

The problem I tackle here does not affect digest calculation, it's in an earlier phase. So it's independent of #7620.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4031/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4033/

tgrabiec commented:
In V2:
  - fixed a bug where a combined change which contains an earlier deletion of the keyspace, creation of the keyspace, and creation of a table would not create the table; The final set affected tables should be a union of existing tables and affected tables from the mutations; Unit test was added for this;
  - introduce more types to express the meaning of objects.
  - Split affected tables into tables and views to avoid reading schema_mutations from system_schema.views for each table which we know is not a view, and vice versa

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4047/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4048/

denesb commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1166,88 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
-        keyspaces.emplace(value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0))));
+        sstring keyspace_name = value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0)));
+
+        if (schema_tables_holding_schema_mutations().contains(mutation.schema()->id())) {
```
Do we expect mutations not relevant to the schema here?

tgrabiec replied:
"schema mutations" here refers to the `schema_mutations` object.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1166,88 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+                      m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name);
+        result.all_in_keyspace = true;
+    }
+    for (auto&& e : m.partition().row_tombstones()) {
```
Let's rename row_tombstones() one day.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1166,88 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+            break;
+        }
+        auto table_name = get_table_name(rt.start);
+        if (table_name != get_table_name(rt.end)) {
```
Minor, but I think folding the table comparison into the previous `if` would make the code a little clearer. Similar to testing `!some_optional || *some_optional != ...`.

avikivity replied:
It would double increase the calls to get_table_name() through.

avikivity commented:
Alternative: load the mutations once during boot, and keep them in memory always.

tgrabiec commented:
> Alternative: load the mutations once during boot, and keep them in memory always.

This is orthogonal. In the scenario I tested, all schema tables are served from cache, but due to fine-grained nature of the algorithm the CPU overhead was significant. This patch reduces the number of unnecessary queries to the schema tables.

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -1095,26 +1166,88 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+    const schema& s = *m.schema();
+    auto get_table_name = [&] (const clustering_key& ck) {
+        // The first component of the clustering key in each table listed in
+        // schema_tables_holding_schema_mutations contains the table name.
```
One would have to add this new table to `schema_tables_holding_schema_mutations` explicitly, so it would have to involve an intentional mistake on his part.

You want to have a unit test which walks over tables `schema_tables_holding_schema_mutations` and checks if the clustering column is called "table_name"?

avikivity replied:
Adding to schema_tables_holding_schema_mutations doesn't mean the adder knows there must be a clustering key called table_name. So let's check it.

tgrabiec commented:
In V3:
 - added asserts to schema_tables_holding_schema_mutations() to verify that all tables have a "table_name" as the first clustering column

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4402/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4431/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4458/

tgrabiec commented:
In V4:
  - Handle legacy schema tables; Fixes assertion failure when legacy schema migrator is invoked, in partitioner_tests.py::TestPartitioner::test_murmur3_partitioner

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4523/
